### PR TITLE
feat: implement custom role management with CRD support

### LINKED ISSUE #40: User Story: Exploring CRD Integration with Kubernetes Default Roles and Keycloak in kube-oidc-proxy
**As a** platform administrator,
**I want to** explore the possibility of using a Custom Resource Definition (CRD) to map Kubernetes default roles in `kube-oidc-proxy`,
**So that** role-based access control (RBAC) can be dynamically managed while leveraging built-in Kubernetes roles and Keycloak for authentication.

**Exploration Goals:**
1. Identify how Kubernetes default roles (e.g., `admin`, `edit`, `view`) can be referenced within a CRD.
2. Determine the feasibility of integrating CRD-based role definitions with `kube-oidc-proxy` authentication and Keycloak.
3. Explore mechanisms to sync CRD-defined roles with Kubernetes RBAC policies dynamically.
4. Evaluate how Keycloak roles and groups can be mapped to Kubernetes default roles via CRD.
5. Test different use cases to validate access control behavior using this approach.

**Expected Outcome:**
- A clear understanding of how a CRD can be designed to use Kubernetes default roles with Keycloak authentication.
- Identification of potential limitations or conflicts with existing Kubernetes RBAC and Keycloak role mappings.
- Recommendations for implementation or alternative approaches if necessary.

### PR DESCRIPTION
**Pull request by husenkuresh1**
This PR introduces a comprehensive solution for managing RBAC configurations dynamically across multiple clusters using Custom Resource Definitions (CRDs). The changes enable centralized control over Kubernetes RBAC resources, simplifying role and binding management in multi-cluster environments.

## Fixes Issue
Closes #40 

#### 1. CustomRole CRD Integration:
- Introduces a `CustomRole` CRD (`custom-rbac.improwised.com/v1`) to declaratively define RBAC roles and bindings.

#### 2. Dynamic RBAC Reconciliation
- Adds a `CustomRoleWatcher` that monitors `CustomRole` resources using Kubernetes informers.
- Processes CRD changes (`add` / `update` / `delete`) to dynamically update RBAC configurations across specified clusters.
- Handles cache synchronization and reconciliation of existing resources during startup.

#### 3. Multi-Cluster Support
- Allows targeting specific clusters or applying configurations globally using `clusters` fields in CRDs.
- Supports namespaced and cluster-wide bindings for granular access control.

---

### New Components

- `constants/main.go`: Defines CRD group/version/resource constants.
- `crd/customrole_processor.go`: Converts CRDs to Kubernetes RBAC resources and applies them to clusters.
- `crd/customrole_watcher.go`: Implements informer logic and event handlers for CRD lifecycle management.

---

### Enhanced Testing

- Adds extensive unit tests (`customrole_processor_test.go`) covering role/binding creation, error handling, and edge cases.
- Validates CRD-to-RBAC conversions, cluster targeting, and namespace filtering.

---

### 🛠 Miscellaneous Improvements

- Adds copyright headers to critical files.
- Updates `go.mod` with testing dependencies (`stretchr/testify`).
- Fixes code formatting and indentation inconsistencies.

---

## Check List (Check all the applicable boxes) 

- [x] My code follows the code style of this project.
- [x] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

### CODE REVIEW

File: constants/main.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Copyright Jetstack Ltd. See LICENSE for details.
+
+package constants
+
+const (
+	CRDGroup    = "custom-rbac.improwised.com"
+	CRDVersion  = "v1"
+	CRDResource = "customroles"
+)
```
Comment (deltasquare4) [inline constants/main.go:9]: We should rename them to more fitting names.

**Review by deltasquare4 — CHANGES_REQUESTED** [review]
We may have to take a look at proxy tests. Other than that and the comments, looks good.
